File: tzio-core/scala/src/main/scala/com/kineolyan/tzio/v1/scala/runner/SystemExecutor.scala
package com.kineolyan.tzio.v1.scala.runner

import java.io.{InputStream, PrintStream}
import java.util.Scanner
import java.util.concurrent.LinkedBlockingDeque

import com.kineolyan.tzio.v1.scala.env.ScalaTzEnv

/**
 * Implementation of the executor based on streams.
 */
class SystemExecutor(in: InputStream, out: PrintStream) {

	def run(env: ScalaTzEnv): Unit = {
		val inputs = new LinkedBlockingDeque[Array[Int]]()
		val errors = new LinkedBlockingDeque[Throwable]()
		val inputThread = new Thread(
			() => {
				val scanner = new Scanner(in)
				while (scanner.hasNextLine) {
					val input = scanner.nextLine()
						.split("\\s*" + SystemExecutor.SPLIT_CHAR + "\\s*")
						.map(c => c.toInt)
					inputs.offer(input)
				}
			},
			"input-thread")
		inputThread.setDaemon(true)
		inputThread.setUncaughtExceptionHandler((t: Thread, err: Throwable) => errors.offer(err))
		inputThread.start()

		out.println(s"System up. Waiting for inputs (${env.slots.inputs.length}):")

		var changedEnv = env
		while (errors.peek() == null) {
			// Look for entries
			val input = inputs.poll()
			if (input != null) {
				changedEnv = changedEnv.consume(input)
			}

			val envAfterTick = changedEnv.tick()
			val result = envAfterTick.collect()
			changedEnv = result._1
			result._2 match {
				case Some(output) =>
					val values = output.map(v => if (v.isPresent) v.getAsInt.toString else "")
						.mkString(SystemExecutor.SPLIT_CHAR)
					out.print("> ")
					out.println(values)
				case None => // Do nothing
			}
		}

		if (!errors.isEmpty) {
			val failure = new RuntimeException("Failure while processing inputs")
			errors.forEach(e => failure.addSuppressed(e))
			throw failure
		}
	}
}

object SystemExecutor {

	/** Input and output value separator */
	private def SPLIT_CHAR: String = ";"

	/**
		* Creates a new instance from the system.
		* @return the created instance
		*/
	def fromSystem(): SystemExecutor =
		new SystemExecutor(System.in, System.out)

}

File: tzio-core/java/src/main/java/com/kineolyan/tzio/v1/java/execs/SystemExecutor.java
package com.kineolyan.tzio.v1.java.execs;

import com.kineolyan.tzio.v1.java.JavaTzEnv;
import java.io.InputStream;
import java.io.PrintStream;
import java.util.OptionalInt;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.TimeUnit;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;

/**
 * Implementation of the executor based on streams.
 */
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class SystemExecutor implements TzExecutor {

	/**
	 * Input stream of this executor
	 */
	private final InputStream in;
	/**
	 * Output stream of this executor
	 */
	private final PrintStream out;

	/**
	 * Creates a new instance from the system.
	 *
	 * @return the created instance
	 */
	public static SystemExecutor fromSystem() {
		return new SystemExecutor(System.in, System.out);
	}

	@Override
	public void run(final JavaTzEnv env) {
		final var outputPrinter = new OutputToStreamCollector(this.out);
		env.produceInto(outputPrinter);

		final BlockingDeque<OptionalInt[]> inputs = new LinkedBlockingDeque<>();
		final BlockingDeque<Throwable> errors = new LinkedBlockingDeque<>();
		final Thread inputThread = createInputThread(inputs, errors);
		inputThread.start();

		try {
			this.out.printf("System up. Waiting for inputs (%d):%n", env.getInputCount());
			while (errors.peek() == null) {
				// Look for entries
				final OptionalInt[] input;
				input = inputs.poll(10, TimeUnit.MILLISECONDS);
				if (input != null) {
					env.consume(input);
				}

				env.tick();
			}
		} catch (InterruptedException e) {
			throw new RuntimeException("Executor interrupted", e);
		}

		final RuntimeException failure = new RuntimeException("Failure while processing inputs");
		errors.forEach(failure::addSuppressed);
		throw failure;
	}

	/**
	 * Creates a thread (not started) parsing the inputs from {@link #in}.
	 *
	 * @param inputs queue filled with the parsed inputs
	 * @param errors errors reported while parsing
	 * @return the thread executing input reading
	 */
	private Thread createInputThread(
			final BlockingDeque<OptionalInt[]> inputs,
			final BlockingDeque<Throwable> errors) {
		final Thread inputThread = new Thread(
				new InputFromStreamCollector(this.in, inputs),
				"input-thread");
		inputThread.setDaemon(true);
		inputThread.setUncaughtExceptionHandler((t, err) -> errors.offer(err));
		return inputThread;
	}
}
